crates/apps/src/lib/node/ledger/shell/finalize_block.rs

```diff
@@ -652,6 +503,372 @@ where
+                .extend(ValidMaspTx((tx_data.tx_index, None)));
+        }
+
+        match dispatch_result {
+            Ok(()) => self.handle_inner_tx_results(
+                response,
+                tx_result,
+                tx_data,
+                &mut tx_logs,
+            ),
+            Err(Error::TxApply(protocol::Error::WrapperRunnerError(msg))) => {
+                tracing::info!(
+                    "Wrapper transaction {} failed with: {}",
+                    tx_logs.tx_event["hash"],
+                    msg,
+                );
+                tx_logs
+                    .tx_event
+                    .extend(WithGasUsed(tx_data.consumed_gas))
+                    .extend(Info(msg.to_string()))
+                    .extend(Code(ResultCode::InvalidTx));
+                // Make sure to clean the write logs for the next transaction
+                self.state.write_log_mut().drop_tx();
+            }
+            Err(msg) => {
+                // This branch represents an error that affects the entire
+                // batch
+                tracing::info!(
+                    "Transaction {} failed with: {}",
+                    tx_logs.tx_event["hash"],
+                    msg
+                );
+
+                tx_logs
+                    .tx_event
+                    .extend(WithGasUsed(tx_data.consumed_gas))
+                    .extend(Info(msg.to_string()))
+                    .extend(Code(ResultCode::WasmRuntimeError));
+
+                self.handle_batch_error(
+                    response,
+                    msg,
+                    tx_result,
+                    tx_data,
+                    &mut tx_logs,
+                );
+            }
+        }
+    }
+
+    // Evaluate the results of all the transactions of the batch. Commit or drop
+    // the storage changes, update stats and event, manage replay protection.
+    fn handle_inner_tx_results(
+        &mut self,
+        response: &mut shim::response::FinalizeBlock,
+        tx_result: namada::tx::data::TxResult<protocol::Error>,
+        tx_data: TxData,
+        tx_logs: &mut TxLogs,
+    ) {
+        // Track the need to commit the batch hash for replay protection. Hash
+        // must be written if at least one of the txs in the batch requires so
+        let mut commit_batch_hash = false;
+        // Track if any of the inner txs failed or was rejected
+        let mut is_any_tx_invalid = false;
+        let mut temp_log = TempTxLogs::default();
+
+        temp_log.check_inner_results(
+            &tx_result,
+            tx_data.tx_header,
+            tx_data.tx_index,
+            tx_data.height,
+            &mut commit_batch_hash,
+            &mut is_any_tx_invalid,
+        );
+
+        if tx_data.is_atomic_batch && is_any_tx_invalid {
+            // Atomic batches need custom handling when even a single tx fails,
+            // since we need to drop everything
+            let unrun_txs = tx_data.commitments_len
```

Comment:
Why do we need to do this subtraction? I don't quite follow.

Reply:
Some transactions of the batch might not be run (because we run out of gas before or if it is an atomic batch because one of the previous txs failed). Here we get how many inner txs fall under this category so that we can update the stats correctly.

Actually about this,

Reply:
I see, understood.